### ISSUE TITLE
Cargo run failing with Cosmos SDK v0.40.0-rc5

### ISSUE THREAD

**Issue by andynog** [cosmos/cosmos-rust#23]
In `ibc-proto` we've been experiencing an error when compiling the proto files. We intend to migrate our proto crate to `cosmos-rust `in the future but just as a test we've tried to reproduce the same error in `cosmos-rust` and it seems to be affecting it too.

### How to reproduce

If you set the Cosmos SDK to release `v0.40.0-rc5` the error below is happening. I've tried to troubleshoot it and see if I could fix but without success so far. So just reporting it here to see if you can help find a solution.

in `proto-build -> src -> main.rs` I've changed this line to point to the Cosmos SDK mentioned above:

```
/// The Cosmos commit or tag to be cloned and used to build the proto files
const COSMOS_REV: &str = "v0.40.0-rc5";
```

Then running it I get an error:

```
$ cargo run
   Compiling proto-build v0.1.0 (/home/andy/development/github.com/cosmos/cosmos-rust/proto-build)
    Finished dev [unoptimized + debuginfo] target(s) in 2.62s
     Running `target/debug/proto-build`
[info] Updating cosmos/cosmos-sdk submodule...
HEAD is now at 7c1da3d99 update changelog for v0.40.0-rc5 (#8156)
[info] Compiling .proto files to Rust into '/tmp/tmp-protobuf/'...
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Custom { kind: Other, error: "protoc failed: /home/andy/development/github.com/cosmos/cosmos-rust/proto-build/../proto: warning: directory does not exist.\nibc/core/client/v1/tx.proto:8:1: warning: Import ibc/core/client/v1/client.proto is unused.\ncosmos/tx/v1beta1/service.proto:7:1: warning: Import gogoproto/gogo.proto is unused.\ncosmos/base/tendermint/v1beta1/query.proto:87:3: \"tendermint.types.BlockID\" is resolved to \"cosmos.base.tendermint.types.BlockID\", which is not defined. The innermost scope is searched first in name resolution. Consider using a leading \'.\'(i.e., \".tendermint.types.BlockID\") to start from the outermost scope.\ncosmos/base/tendermint/v1beta1/query.proto:88:3: \"tendermint.types.Block\" is resolved to \"cosmos.base.tendermint.types.Block\", which is not defined. The innermost scope is searched first in name resolution. Consider using a leading \'.\'(i.e., \".tendermint.types.Block\") to start from the outermost scope.\ncosmos/base/tendermint/v1beta1/query.proto:96:3: \"tendermint.types.BlockID\" is resolved to \"cosmos.base.tendermint.types.BlockID\", which is not defined. The innermost scope is searched first in name resolution. Consider using a leading \'.\'(i.e., \".tendermint.types.BlockID\") to start from the outermost scope.\ncosmos/base/tendermint/v1beta1/query.proto:97:3: \"tendermint.types.Block\" is resolved to \"cosmos.base.tendermint.types.Block\", which is not defined. The innermost scope is searched first in name resolution. Consider using a leading \'.\'(i.e., \".tendermint.types.Block\") to start from the outermost scope.\ncosmos/base/tendermint/v1beta1/query.proto:113:3: \"tendermint.p2p.DefaultNodeInfo\" is resolved to \"cosmos.base.tendermint.p2p.DefaultNodeInfo\", which is not defined. The innermost scope is searched first in name resolution. Consider using a leading \'.\'(i.e., \".tendermint.p2p.DefaultNodeInfo\") to start from the outermost scope.\ncosmos/base/tendermint/v1beta1/query.proto:4:1: warning: Import gogoproto/gogo.proto is unused.\ncosmos/base/tendermint/v1beta1/query.proto:6:1: warning: Import google/api/annotations.proto is unused.\ncosmos/base/tendermint/v1beta1/query.proto:9:1: warning: Import tendermint/types/types.proto is unused.\ncosmos/base/tendermint/v1beta1/query.proto:8:1: warning: Import tendermint/types/block.proto is unused.\ncosmos/base/tendermint/v1beta1/query.proto:7:1: warning: Import tendermint/p2p/types.proto is unused.\n" }', proto-build/src/main.rs:127:47
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Seems something broken regarding the external reference to tendermint protos. I've tried to play with different the `config.extern_path(".tendermint", "::tendermint_proto");` line but no lucky. The prost (and tonic) docs are not very straightforward about this function.

**Comment by tony-iqlusion** [comment]
Confirm that I was able to reproduce this problem.

Here's a more nicely formatted version of the `protoc` errors:

```
protoc failed: /Users/bascule/src/cosmos-rust/proto-build/../proto: warning: directory does not exist.
ibc/core/client/v1/tx.proto:8:1: warning: Import ibc/core/client/v1/client.proto is unused.
cosmos/tx/v1beta1/service.proto:7:1: warning: Import gogoproto/gogo.proto is unused.
cosmos/base/tendermint/v1beta1/query.proto:87:3: "tendermint.types.BlockID" is resolved to "cosmos.base.tendermint.types.BlockID", which is not defined. The innermost scope is searched first in name resolution. Consider using a leading '.'(i.e., ".tendermint.types.BlockID") to start from the outermost scope.
cosmos/base/tendermint/v1beta1/query.proto:88:3: "tendermint.types.Block" is resolved to "cosmos.base.tendermint.types.Block", which is not defined. The innermost scope is searched first in name resolution. Consider using a leading '.'(i.e., ".tendermint.types.Block") to start from the outermost scope.
cosmos/base/tendermint/v1beta1/query.proto:96:3: "tendermint.types.BlockID" is resolved to "cosmos.base.tendermint.types.BlockID", which is not defined. The innermost scope is searched first in name resolution. Consider using a leading '.'(i.e., ".tendermint.types.BlockID") to start from the outermost scope.
cosmos/base/tendermint/v1beta1/query.proto:97:3: "tendermint.types.Block" is resolved to "cosmos.base.tendermint.types.Block", which is not defined. The innermost scope is searched first in name resolution. Consider using a leading '.'(i.e., ".tendermint.types.Block") to start from the outermost scope.
cosmos/base/tendermint/v1beta1/query.proto:113:3: "tendermint.p2p.DefaultNodeInfo" is resolved to "cosmos.base.tendermint.p2p.DefaultNodeInfo", which is not defined. The innermost scope is searched first in name resolution. Consider using a leading '.'(i.e., ".tendermint.p2p.DefaultNodeInfo") to start from the outermost scope.
cosmos/base/tendermint/v1beta1/query.proto:4:1: warning: Import gogoproto/gogo.proto is unused.
cosmos/base/tendermint/v1beta1/query.proto:6:1: warning: Import google/api/annotations.proto is unused.
cosmos/base/tendermint/v1beta1/query.proto:8:1: warning: Import tendermint/types/block.proto is unused.
cosmos/base/tendermint/v1beta1/query.proto:9:1: warning: Import tendermint/types/types.proto is unused.
cosmos/base/tendermint/v1beta1/query.proto:7:1: warning: Import tendermint/p2p/types.proto is unused.
```

**Comment by jkilpatr** [comment]
So the problem here is that in order to pass through some ABCI stuff for RC5 tindermint proto types where pulled into Cosmos-sdk proto definitions, the proto compiler is looking for one of these files 

https://github.com/tendermint/tendermint/tree/master/proto/tendermint

and cant' find it, our compiler needs to be updated to keep them in scope (note that this may remove our dependency on tendermint-proto the crate, since we need to compile those protos ourselves anyways.

**Comment by tony-iqlusion** [comment]
I think this is actually an issue with the upstream proto definitions coming from `cosmos/cosmos-sdk`. You can (inadvertently) see the same error in this (otherwise unrelated) issue:

https://github.com/cosmos/cosmos-sdk/issues/8171

Adding a submodule for `tendermint/tendermint` doesn't help: as you can see in the protoc output from above, these definitions are already available, and unused:

```
cosmos/base/tendermint/v1beta1/query.proto:8:1: warning: Import tendermint/types/block.proto is unused.
cosmos/base/tendermint/v1beta1/query.proto:9:1: warning: Import tendermint/types/types.proto is unused.
cosmos/base/tendermint/v1beta1/query.proto:7:1: warning: Import tendermint/p2p/types.proto is unused.
```

The real problem appears to be an upstream scoping issue. You can see the error here:

> "tendermint.types.BlockID" is resolved to "cosmos.base.tendermint.types.BlockID", which is not defined. The innermost scope is searched first in name resolution. Consider using a leading '.'(i.e., ".tendermint.types.BlockID") to start from the outermost scope.

There are two proto packages named `tendermint`, the toplevel `.tendermint` package and `cosmos.base.tendermint`. The protos where this error is occurring is inside of the `cosmos.base.tendermint.v1beta1` package, and so "tendermint" is resolving to `cosmos.base.tendermint` instead of `.tendermint`.

Not sure why this isn't breaking the upstream build as per the linked issue this does still seem to be causing (potentially ignored) upstream errors. That said...

Adding a leading `.` to the type names as suggested by protoc corrects the issue, but requires an upstream change. I can attempt a PR.

**Comment by tony-iqlusion** [comment]
Opened a PR to fix the scoping here: https://github.com/cosmos/cosmos-sdk/pull/8236

**Comment by tony-iqlusion** [comment]
https://github.com/cosmos/cosmos-sdk/pull/8236 has been merged! 🎉 

I can also confirm that if I bump `COSMOS_REV` to the latest commit locally, everything is working again:

```
[info] Updating cosmos/cosmos-sdk submodule...
HEAD is now at cfeb5eb4d Fix tendermint query proto scoping (#8236)
[info] Compiling .proto files to Rust into '/tmp/tmp-protobuf/'...
[info ] Compiling proto clients for GRPC services!
[info ] => Done!
[info ] Copying generated files into '../cosmos-sdk-proto/src/prost/'...
```

Will go ahead and close this issue out as it appears we're free to bump `COSMOS_REV` as soon as e.g. v0.40.0-rc6 gets tagged.
